### DISCUSSION TITLE
New Savegame Format Features

### DISCUSSION THREAD

**Discussion by walkawayy** [LostArtefacts/TRX#417] [Q&A] [answered]
Now that a new savegame format has been introduced, let's tackle a list of issues and feature requests that can be implemented before backwards compatibility will be broken.

- [x] #277
- [x] #170 (could be nice to add this in same release as other save improvements)
- [x] #406
- [x] #87
- [x] #365
- [x] #396
- [x] #48
- [x] ~~#197~~
- [x] #336
- [x] #371
- [x] #388
- [x] #385
- [x] ~~#226~~
- [x] #418

**Comment by walkawayy** [comment]
@rr- let me know what you think. Tried to gather all issues related to savegames. Also I'm pretty busy until Friday, so I won't be able to provide much testing / features until then. I can try and finish / test restart level at least. But I don't think we need to rush this since this is a big change. Maybe it should be a 3.0 release?

**Reply by rr-** [reply]
Hmm I don't think it's big enough to be 3.0, the users won't notice much difference TBH.
I looked at all the issues and apart from #336 I think the only things missing is saving the last played music track and death count.

**Reply by rr-** [reply]
3.0 would maybe make sense if we implemented both level select and restart level in that same release, but I'm not a fan of this and prefer to RERO (as in [this](https://en.wikipedia.org/wiki/Release_early,_release_often), not [this](https://knowyourmeme.com/memes/rero-rero)). I feel that waiting for these two features would considerably delay releasing what we have so far, and we have a lot of things already. At the same time, I think we've got everything ready for both of them, but I might be missing something.

**Reply by walkawayy** [reply]
Numbering for easier discussion:

1. I think restart level can definitely be included in this release. I think it's ready now that Lara's starting health is stored in the save games for the disable healing feature. For legacy saves, it can just restart at full HP which will be a rare use case since any future level saves will have that starting HP stored.

2. Select level is going to require a new menu option somewhere. On PS1, you can't save anywhere so select level is in place of load game. I was thinking the unused Natla Map could be used as a new menu option, but it's included in the level phds and not the Title.phd. As long as the backend for storing each level's starting items and stats is in place, then I agree it's fine to push it to a later release. Same thing with game completion stats.

3. Are particle saving and music trigger saving ready? Maybe I missed that.

**Reply by rr-** [reply]
3. No they need to be coded. Would you like to take these on?

**Reply by walkawayy** [reply]
It depends on how soon you want to release. I can't promise anything before the weekend.

Another thought I had that Arsunt mentioned a while ago, is it possible to save current music track info? For example, if I trigger the T Rex music, save and reload, could it resume where it left off? Seems like a super niche use case that might not be very important tho.

**Reply by rr-** [reply]
We can save anything we want now. Even the entire level geometry :D question is how feasible it would be to integrate with the rest of the codebase. For music I think it's easy, just the track, and current buffer position.
I'd like to release some time early next week or this weekend.

**Reply by walkawayy** [reply]
Ok I striked out level select out but the backend is in place for it to be added in the future.

I'll assign issues to myself and add 2.6 milestones for:
1. Restart level (basically done)
2. Prevent music triggers from playing again
3. Emitters
4. Possibly saving current music track / buffer position

Does that work? Seems like the rest of the 2.6 issues just need PR testing. Only thing I think is left undone / unassigned would be saving death count and adding that to level stats and end game stats?

**Reply by rr-** [reply]
Yeah. I can take care of that before the weekend :+1:

**Reply by Richard-L** [reply]
@walkawayy 
- **Select level**: can you add a new page after "Load Game" in the passport? That would make most sense to me, as it's a PC feature next to a PS1/Saturn one. I didn't know the passport had that many textures, but it seems there are a few options 
![image](https://user-images.githubusercontent.com/4934209/153081472-ac4b20ed-63e5-4ab7-ae0d-965696577527.png)
^ where are the later passport pages with some travel stamps in used? This [texture dump](https://cdn.discordapp.com/attachments/309026200284823552/936840064179249162/TR1_textures.png) is from XProger.
- **Saving the music buffer** and enabling an adequate "music position resume upon load" would be huge imo.

Side notes:
- I love the RERO wikipedia link, even if it's not my nature.
- More people are asking re gamepad support in the Discord now. Perhaps something to consider before a big new version number?

**Reply by rr-** [reply]
The stamps are used to texture the first page depending on the progress of the game.
Sadly we cannot alter the passport number of pages without hacking the mesh data.

**Reply by Richard-L** [reply]
Oh the progress indication is so sweet. Good game.

Shame about the passport indeed.

**Comment by Rapora9** [comment]
Two issues that might be relevant:
- unlike normal switches, underwater switches can be pulled again even if they're set as one-shot. I haven't found them causing any issues in TR1 (there are rarely one-shot underwater switches anyway), but they caused a problem in a custom level where the switch was not activating all flags/bits but only 1.
- A common speedrun strat: in Sanctuary of the Scion, an underwater current takes Lara away when the rooms are flipped after using an underwater switch under the huge statue. However, if player saves and loads just after using the lever, the current doesn't work. Why? Indication of some bigger problem with flipmaps/currents?

**Reply by Richard-L** [reply]
Can you link a video of the speedrun strat? I don't remember it.

**Reply by Rapora9** [reply]
Using Footi as a source again: https://youtu.be/i6Am1jdLC7M?t=2690

**Comment by rr-** [comment, marked as answer]
I believe this can be closed. @walkawayy
